### PR TITLE
fix: provide function to support hostname mapping

### PR DESCRIPTION
**Pull request by junczhu**
Resolve: #542 
Signed-off-by: Juncheng Zhu <junczhu@microsoft.com>

### CODE REVIEW

**Comment by qweeah** [comment]
What if username/password are passed through flags? Replacing host name in cred store only works when credential is empty,  see 
https://github.com/oras-project/oras/blob/main/cmd/oras/internal/option/remote.go#L152

File: internal/credential/store.go [hunk 1]
```diff
@@ -113,3 +116,11 @@ func (s *Store) Credential(ctx context.Context, registry string) (auth.Credentia
 	}
 	return auth.EmptyCredential, nil
 }
+
+// convertHostname maps docker.io to registry-1.docker.io
+func convertHostname(registry string) string {
```
Comment (qweeah) [inline internal/credential/store.go:121]: use [registry.Host](https://github.com/oras-project/oras-go/blob/main/registry/reference.go#L177) directly

**Comment by shizhMSFT** [comment]
@qweeah I've discussed offline with @junczhu. The behavior of `docker.io` in docker is special that it connects to the host of `registry-1.docker.io` using the credential with the reigstry name `https://index.docker.io/v1/`, which is really not a valid name (but it uses).

**Review by shizhMSFT — CHANGES_REQUESTED** [review]
Please update this PR as discussed to use the credentials of `https://index.docker.io/v1/` for `docker.io` when connecting `registry-1.docker.io` as `docker.io`. Meanwhile, please well document the docker behavior in the corresponding comment block.

File: cmd/oras/internal/option/remote.go [hunk 2]
```diff
@@ -110,7 +110,7 @@ func (opts *Remote) tlsConfig() (*tls.Config, error) {
 }
 
 // authClient assembles a oras auth client.
-func (opts *Remote) authClient(debug bool) (client *auth.Client, err error) {
+func (opts *Remote) authClient(host string, debug bool) (client *auth.Client, err error) {
```
Comment (qweeah) [inline cmd/oras/internal/option/remote.go:113]: Why we need this new argument? It's not used anywhere.

**Review by qweeah — CHANGES_REQUESTED** [review]
Can you add a unit test case for this host name replacement?

File: cmd/oras/login.go [hunk 3]
```diff
@@ -116,7 +116,14 @@ func runLogin(opts loginOptions) (err error) {
 	if err != nil {
 		return err
 	}
-	if err := store.Store(opts.Hostname, opts.Credential()); err != nil {
+	// For a user case that login 'docker.io',
+	// According the the behavior of Docker CLI,
+	// credential should be added under key "https://index.docker.io/v1/"
+	hostname := opts.Hostname
+	if opts.Hostname == "docker.io" {
```
Comment (shizhMSFT) [inline cmd/oras/login.go:123]:
```suggestion
	if hostname  == "docker.io" {
```

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**